lib/api/core/models/repositories/repository.js

```diff
@@ -1,344 +1,421 @@
-
-/**
- * Initialize the repository
- *
- * @param options
- */
-Repository.prototype.init = function repositoryInit (options) {
-  // if set to null, it means that we explicitely dont want a databaseEngine for this repository
-  // as null is falsy, without this check, the databaseEngine is always set...
-  if (options.databaseEngine === null) {
-    this.databaseEngine = options.databaseEngine;
-  }
-  else {
-    this.databaseEngine = options.databaseEngine || this.kuzzle.internalEngine;
-  }
-  if (options.cacheEngine === null) {
-    this.cacheEngine = options.cacheEngine;
+class Repository {
+  /**
+   * @param {Kuzzle} kuzzle
+   * @param {string} index
+   * @constructor
+   */
+  constructor(kuzzle, index = kuzzle.internalEngine.index) {
+    this.kuzzle = kuzzle;
+    this.index = index;
+    this.ttl = kuzzle.config.repositories.common.cacheTTL;
+    this.collection = null;
+    this.ObjectConstructor = null;
+    this.databaseEngine = null;
+    this.cacheEngine = null;
   }
-  else {
-    this.cacheEngine = options.cacheEngine || this.kuzzle.services.list.internalCache;
+
+  /**
+   * Initialize the repository
+   *
+   * @param {object} options
+   */
+  init (options) {
+    // if set to null, it means that we explicitely dont want a databaseEngine for this repository
+    // as null is falsy, without this check, the databaseEngine is always set...
+    if (options.databaseEngine === null) {
+      this.databaseEngine = options.databaseEngine;
+    }
+    else {
+      this.databaseEngine = options.databaseEngine || this.kuzzle.internalEngine;
+    }
+    if (options.cacheEngine === null) {
+      this.cacheEngine = options.cacheEngine;
+    }
+    else {
+      this.cacheEngine = options.cacheEngine || this.kuzzle.services.list.internalCache;
+    }
   }
-  this.cacheKeyPrefix = 'repos/' + this.index + '/' + this.collection + '/';
-};
 
-/**
- *
- * @param {string} id
- * @returns {Promise} resolves on a new ObjectConstructor()
- */
-Repository.prototype.loadOneFromDatabase = function repositoryLoadOneFromDatabase (id) {
-  var result;
-
-  return this.databaseEngine.get(this.collection, id)
-    .then(response => {
-      if (response._id) {
-        result = new this.ObjectConstructor();
-        if (response._source) {
-          return _.assignIn(result, response._source, {_id: response._id});
-        }
-        return _.assignIn(result, response);
-      }
-      return null;
-    })
-    .catch(error => {
-      if (error.status === 404) {
-        // no content found, we return null without failing
-        return Promise.resolve(null);
-      }
+  /**
+   * @param {string} id
+   * @param {string} [collection]
+   * @returns {Promise} resolves on a new ObjectConstructor()
+   */
+  loadOneFromDatabase (id, collection = this.collection) {
+    let result;
 
-      return Promise.reject(error);
-    });
-};
+    if (collection === null) {
+      return Promise.reject(new InternalError('The collection must be specified.'));
```

Comment:
It can not happen due to a bad request, it might come from an bad plugin implementation, will use this one instead.